### PR TITLE
Add ability to pass arguments to generator function in external_source op

### PR DESCRIPTION
**Pull request by szalpal**
Signed-off-by: Michał Szołucha <mszolucha@nvidia.com>

#### Why we need this PR?
*Pick one, remove the rest*

- It adds new feature: Ability to provide argument list to a generator function in external_source op.

The need for this feature emerged, when I was migrating from old object-API to the new fn-API. Previously,
since the Pipeline class had a state, it was possible to use member variables as "arguments" in the generator function.
With the fn API we'd need to use global variables. However, using an additional argument to `external_source` operator, we can pass the arguments to the generator function within the `external_source`.

#### What happened in this PR?
*Fill relevant points, put NA otherwise. Replace anything inside []*
 - What solution was applied:
     Added an argument
 - Affected modules and functionalities:
     Python layer in the external_source operator
 - Key points relevant for the review:
     NA
 - Validation and testing:
     Yes
 - Documentation (including examples):
     Yes

**JIRA TASK**: *[Use DALI-XXXX or NA]*

### CODE REVIEW

**Comment by szalpal** [comment]
!build

File: dali/python/nvidia/dali/external_source.py [hunk 1]
```diff
@@ -364,6 +371,11 @@ class ExternalSource():
     ``iter(source)`` returns a fresh iterator on each call, or a generator function.
     In the latter case, the generator function is called again when more data than was
     yielded by the function is requested.
+    
+    Please note: when the ``source`` is a generator function, and the cycling is turned on (either
```
Comment (mzient) [inline dali/python/nvidia/dali/external_source.py:375]:
```suggestion
    Please note that when the ``source`` is a generator function, and the cycling is turned on (either
```

File: dali/python/nvidia/dali/external_source.py [hunk 2]
```diff
@@ -67,25 +67,28 @@ def __next__(self):
...
+
+class _CycleGenFunc:
+    def __init__(self, gen_func, mode, gen_func_args={}):
         self.source = gen_func
+        self.gen_func_args = gen_func_args
         self.signaling = (mode == "raise")
 
     def __iter__(self):
-        self.it = iter(self.source())
+        self.it = iter(self.source(**self.gen_func_args))
```
Comment (mzient) [inline dali/python/nvidia/dali/external_source.py:78]: This will work only with keyword arguments.

**Review by mzient — CHANGES_REQUESTED** [review]
I think that this feature is introduced to satisfy a particular need without considering its rather broad impact on External Source API.
1. It's incomplete - it works only with keyword arguments.
2. It gives a bit too much of special treatment for generator functions
3. It's redundant. This can be achieved by simply doing this:
```Python
inp = fn.external_source(source=lambda: generator(arg1, arg2, kwarg=...))
```

**Pull request closed without merging**